blocklib/digitizers/lib/digitizer_block_impl.cc

```diff
@@ -729,41 +717,40 @@ work_return_t digitizer_block_impl::work_rapid_block(work_io& wio)
             return work_return_t::ERROR; // TODO(PORT) was -1
         }
 
-        // Attach trigger info to value outputs and to all ports
-        auto vec_idx = 0;
-        uint32_t pre_trigger_samples_with_downsampling =
-            get_pre_trigger_samples_with_downsampling();
-        double time_per_sample_with_downsampling_ns =
-            d_time_per_sample_ns * d_downsampling_factor;
-
-        for (auto i = 0; i < d_ai_channels && vec_idx < (int)wio.outputs().size();
-             i++, vec_idx += 2) {
-            if (!d_channel_settings[i].enabled) {
-                continue;
+        if (!d_timing_messages.empty()) {
+            const auto timing = d_timing_messages.front();
+
+            // Attach trigger info to value outputs and to all ports
+            auto vec_idx = 0;
+            const uint32_t pre_trigger_samples_with_downsampling =
+                get_pre_trigger_samples_with_downsampling();
+            const double time_per_sample_with_downsampling_ns =
+                d_time_per_sample_ns * d_downsampling_factor;
+
+            // TODO do we need this timestamp adjustment for pre_samples?
```

Comment:
> > // TODO do we need this timestamp adjustment for pre_samples?
> 
> When you take the timestamp directly from the timing (timing.timestamp), you should not adjust the timestamp in any way. Just place it on the correct sample (tag.offset).
> 
> Adjustment was only needed before to calculate the trigger-stamp by using the latest 'timestamp_now_ns_utc' and the number of pre-trigger samples. Since you instead rely on being in sync with the incoming timing-messages, you should not care about these local utc stamps any more (at least not for trigger-tags, when the trigger-source is 'timing').

Thanks, that's what I thought, so I remove the TODO. The local timestamp is now only used by some acqinfo tag logic, probably that should also be replaced, but I think we need to clarify first which of these other tags we want to keep in which form.

Reply:
Humm, I think you removed a bit too much in https://github.com/fair-acc/gr-digitizers/pull/117/commits/62c1f9dd387ab35245ad58db5089e3031dc22b7a  ... while you dont need to add a offset to the timestamp, you still will need to place the tag on the exact sample where the trigger was detected, which is not the first sample of the measurement if you use pre-trigger-samples.

So I think this line, using `pre_trigger_samples_with_downsampling`  has to be kept:

`const auto tag_offset = wio.outputs()[0].nitems_written() + pre_trigger_samples_with_downsampling;`

Reply:
Oups, fixing the pre-trigger handling.